Tests/PackageGraphTests/PackageGraphTests.swift

```diff
@@ -138,7 +138,26 @@ class PackageGraphTests: XCTestCase {
             "/Bar": Package(name: "Bar", dependencies: [.Package(url: "/Foo", majorVersion: 1)]),
         ], root: "/Bar", diagnostics: diagnostics, in: fs)
 
-        XCTAssertEqual(diagnostics.diagnostics[0].localizedDescription, "multiple targets named 'Bar'")
+        XCTAssertEqual(diagnostics.diagnostics[0].localizedDescription, "multiple targets named 'Bar' in: Bar, Foo")
+    }
+
+    func testNestedDuplicateModules() throws {
+        let fs = InMemoryFileSystem(emptyFiles:
+            "/Fourth/Sources/First/source.swift",
+            "/Third/source.swift",
+            "/Second/source.swift",
+            "/First/source.swift"
+        )
+
+        let diagnostics = DiagnosticsEngine()
+        _ = loadMockPackageGraph([
+            "/Fourth": Package(name: "Fourth"),
+            "/Third": Package(name: "Third", dependencies: [.Package(url: "/Fourth", majorVersion: 1)]),
+            "/Second": Package(name: "Second", dependencies: [.Package(url: "/Third", majorVersion: 1)]),
+            "/First": Package(name: "First", dependencies: [.Package(url: "/Second", majorVersion: 1)]),
+        ], root: "/First", diagnostics: diagnostics, in: fs)
+
+        XCTAssertEqual(diagnostics.diagnostics[0].localizedDescription, "multiple targets named 'First' in: First, Fourth")
```

Comment:
I suspect we will have duplicate diagnostics when a similarly named target is present in 3+ packages. Can you add a test for that and see what we get?